File: src/main/scala/com/juanoff/scala/types/SparseMatrixType.scala
package com.juanoff.scala.types

import java.io.{BufferedReader, InputStreamReader}
import scala.jdk.CollectionConverters.*

class SparseMatrixType extends UserType {
  override def typeName(): String = "SparseMatrix"

  override def create(): Any = new SparseMatrix()

  override def clone(obj: Any): Any = {
    obj match {
      case matrix: SparseMatrix => matrix.copy()
      case _ => throw new IllegalArgumentException(s"Expected ${typeName()}")
    }
  }

  override def readValue(in: InputStreamReader): Any = {
    val reader = new BufferedReader(in)
    val line = reader.readLine()
    parseValue(line)
  }

  override def parseValue(ss: String): Any = {
    if (ss == null || ss.trim.isEmpty) {
      throw new IllegalArgumentException("Enter matrix parameters")
    }

    val matrix = new SparseMatrix()
    val parts = ss.split(";")
    parts.foreach { p =>
      val trimmed = p.trim
      if (trimmed.nonEmpty) {
        val vals = trimmed.split(",")
        if (vals.length != 3) {
          throw new IllegalArgumentException(s"Invalid matrix format in part: $trimmed")
        }

        try {
          val x = vals(0).trim.toInt
          val y = vals(1).trim.toInt
          val v = vals(2).trim.toDouble
          matrix.set(x, y, v)
        } catch {
          case _: NumberFormatException =>
            throw new IllegalArgumentException(s"Invalid number format in part: $trimmed")
        }
      }
    }

    matrix
  }

  override def getTypeComparator: Comparator = {
    (o1: Any, o2: Any) => {
      (o1, o2) match {
        case (m1: SparseMatrix, m2: SparseMatrix) =>
          val sizeCmp = Integer.compare(m1.size(), m2.size())
          if (sizeCmp != 0) sizeCmp else m1.sum().compareTo(m2.sum())
        case _ => 0
      }
    }
  }

  override def serialize(obj: Any): String = {
    obj match {
      case matrix: SparseMatrix =>
        val entries = matrix.getEntries
        entries.asScala.map { e => s"${e.x},${e.y},${e.value}" }.mkString(";")
      case _ => ""
    }
  }

  override def deserialize(s: String): Any = parseValue(s)
}

File: src/main/kotlin/com/juanoff/kotlin/types/SparseMatrixType.kt
package com.juanoff.kotlin.types

import java.io.InputStreamReader

class SparseMatrixType : UserType {
    override fun typeName(): String = "SparseMatrix"
    override fun create(): Any = SparseMatrix()

    override fun clone(obj: Any): Any {
        require(obj is SparseMatrix) { "Expected SparseMatrix, got ${obj::class.simpleName}" }
        return obj.copy()
    }

    override fun readValue(reader: InputStreamReader): Any {
        return reader.buffered().use { reader ->
            parseValue(reader.readLine())
        }
    }

    override fun parseValue(ss: String): Any {
        require(ss.isNotBlank()) { "Enter matrix parameters" }

        val matrix = SparseMatrix()
        val parts = ss.split(";")

        parts.forEach { part ->
            val trimmed = part.trim()
            if (trimmed.isNotEmpty()) {
                val vals = trimmed.split(",")
                require(vals.size == 3) { "Invalid matrix format in part: $trimmed" }

                try {
                    val x = vals[0].trim().toInt()
                    val y = vals[1].trim().toInt()
                    val v = vals[2].trim().toDouble()
                    matrix[x, y] = v
                } catch (e: NumberFormatException) {
                    throw IllegalArgumentException("Invalid number format in part: $trimmed", e)
                }
            }
        }

        return matrix
    }

    override fun getTypeComparator(): Comparator = Comparator { o1, o2 ->
        when {
            o1 is SparseMatrix && o2 is SparseMatrix -> {
                val sizeCmp = o1.size().compareTo(o2.size())
                if (sizeCmp != 0) sizeCmp else o1.sum().compareTo(o2.sum())
            }

            else -> 0
        }
    }

    override fun serialize(obj: Any): String {
        require(obj is SparseMatrix) { "Expected SparseMatrix" }
        return obj.getEntries().joinToString(separator = ";") { "${it.x},${it.y},${it.value}" }
    }

    override fun deserialize(s: String): Any = parseValue(s)
}